shared/src/commonMain/kotlin/com/mbta/tid/mbta_app/model/NearbyHierarchy.kt

```diff
@@ -0,0 +1,433 @@
+package com.mbta.tid.mbta_app.model
+
+import com.mbta.tid.mbta_app.model.NearbyHierarchy.DirectionOrHeadsign
+import com.mbta.tid.mbta_app.model.NearbyHierarchy.LineOrRoute
+import com.mbta.tid.mbta_app.model.NearbyHierarchy.NearbyLeaf
+import com.mbta.tid.mbta_app.model.NearbyHierarchy.RouteAtStop
+import com.mbta.tid.mbta_app.model.response.GlobalResponse
+import com.mbta.tid.mbta_app.model.response.PredictionsStreamDataResponse
+import com.mbta.tid.mbta_app.model.response.ScheduleResponse
+import io.github.dellisd.spatialk.geojson.Position
+import io.github.dellisd.spatialk.turf.ExperimentalTurfApi
+import io.github.dellisd.spatialk.turf.distance
+import kotlinx.datetime.Instant
+
+private typealias InnerLayer = Map<DirectionOrHeadsign, NearbyLeaf>
+
+private typealias MiddleLayerData = RouteAtStop<List<Direction>, InnerLayer>
+
+private typealias MiddleLayer = Map<Stop, MiddleLayerData>
+
+private typealias OuterLayer = Map<LineOrRoute, MiddleLayer>
+
+data class NearbyHierarchy(private val data: OuterLayer) {
+    sealed interface LineOrRoute {
+        data class Line(
+            val line: com.mbta.tid.mbta_app.model.Line,
+            val routes: Set<com.mbta.tid.mbta_app.model.Route>
+        ) : LineOrRoute
+
+        data class Route(val route: com.mbta.tid.mbta_app.model.Route) : LineOrRoute
+    }
+
+    data class RouteAtStop<StopData, T>(val stopData: StopData, val data: T) {
+        fun <U> map(f: (T) -> U): RouteAtStop<StopData, U> = RouteAtStop(stopData, f(data))
+    }
+
+    sealed interface DirectionOrHeadsign {
+        data class Direction(val direction: com.mbta.tid.mbta_app.model.Direction) :
+            DirectionOrHeadsign
+
+        data class Headsign(val headsign: String, val route: Route, val line: Line?) :
+            DirectionOrHeadsign
+    }
+
+    data class PartialHierarchy<StopData, T>(
+        val data: Map<LineOrRoute, Map<Stop, RouteAtStop<StopData, T>>>
```

Comment:
The route isn't part of the structure there, `RouteAtStop` is just the only way to attach the `List<Direction>` to the entire `ByDirectionOrHeadsign`. If you think it'd be clearer to just use a `Pair`, I can, and if you can think of a better name, I'd be happy to hear it.

Reply:
Got it. I think it could be clearer to have separate types for the final state from the intermediate state, along the lines of `ByStopData(directions: List<Direction>, data: ByDirectionOrHeadsign)` defined as a class and PartialByStopData(directions: T, data: U)`. That way, when working with the permanent structure, the field names are clearer. 

Taking that thought further - if there is only one source for the directions and it isn't really being combined with anything, could that be part of the fixed hiearchy structure and removed from the partial zipping entirely?

Reply:
If we put the directions into the map key, then merging the stop and its directions from the scheduled data with the stop without directions from the realtime data gets a lot more complicated.